pallets/maintenance-mode/src/lib.rs

```diff
@@ -162,6 +171,34 @@ pub mod pallet {
 		}
 	}
 
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// on_runtime_upgrade is called once at the first block that a runtime upgrade is applied.
+		/// we use it here to optionally trigger maintenance mode if configured to do so.
+		fn on_runtime_upgrade() -> Weight {
+
+			let mut weight: Weight = 0u64.into();
+			weight += T::DbWeight::get().read; // reading auto-enable flag
+			
+			if AutoEnableOnRuntimeUpgrade::<T>::get() {
+
+				weight += T::DbWeight::get().read; // reading current maintenance mode
+				if MaintenanceMode::<T>::get() {
+					log::warn!("Automatically entering maintenance mode upon runtime upgrade");
+					weight += T::DbWeight::get().read; // writing maintenance mode flag
+					MaintenanceMode::<T>::put(true);
+				} else {
+					log::warn!("Maintenance mode already enabled");
+				}
+
+				weight += T::DbWeight::get().write; // writing auto-enable flag
+				AutoEnableOnRuntimeUpgrade::<T>::put(false);
```

Comment:
Here we automatically unset the auto-enable flag. I'm not sure whether that's ideal or not.